Make `linux-keyutils` optional

messense opened the pull request:
`linux-keyutils` [doesn't compile on some 32-bit platforms](https://github.com/landhb/linux-keyutils/issues/8), enable it unconditionally is troublesome.

landhb commented:
Probably a little hastly. 

I've just fixed the issue and published `0.2.3`. I've also added CI/CD for multiple architectures (including big endian and 32 bit ones). 

0.2.3 should update seamlessly with the current keyring release, since it depends on 0.2.*

Running `cargo update` updates both for me:

```
 $ cargo update
    Updating crates.io index
    Updating keyring v2.0.0-rc.3 -> v2.0.1
    Updating linux-keyutils v0.2.2 -> v0.2.3
 ```

messense commented:
Thank @landhb for the quick fix, but I think it's better to make it optional anyway?

landhb commented:
Technically yes, functionally no, since you'll always be building both the `libc` and `bitflags` crates when building `keyring` for Linux.

`keyutils` may actually be a better back-end choice for `maturin`'s `password-storage` feature. Since it is lightweight and more portable (secret-service requires Gnome keyring which isn't present on a lot of Linux systems).

```
├── linux-keyutils v0.2.1
│   ├── bitflags v1.3.2
│   └── libc v0.2.137
└── secret-service v3.0.1
    ├── aes v0.7.5
    │   ├── cfg-if v1.0.0
    │   ├── cipher v0.3.0
    │   │   └── generic-array v0.14.6
    │   │       └── typenum v1.15.0
    │   │       [build-dependencies]
    │   │       └── version_check v0.9.4
    │   ├── cpufeatures v0.2.5
    │   └── opaque-debug v0.3.0
    ├── block-modes v0.8.1
    │   ├── block-padding v0.2.1
    │   └── cipher v0.3.0 (*)
    ├── futures-util v0.3.25
    │   ├── futures-core v0.3.25
    │   ├── futures-macro v0.3.25 (proc-macro)
    │   │   ├── proc-macro2 v1.0.47
    │   │   │   └── unicode-ident v1.0.5
    │   │   ├── quote v1.0.21
    │   │   │   └── proc-macro2 v1.0.47 (*)
    │   │   └── syn v1.0.103
    │   │       ├── proc-macro2 v1.0.47 (*)
    │   │       ├── quote v1.0.21 (*)
    │   │       └── unicode-ident v1.0.5
    │   ├── futures-sink v0.3.25
    │   ├── futures-task v0.3.25
    │   ├── pin-project-lite v0.2.9
    │   ├── pin-utils v0.1.0
    │   └── slab v0.4.7
    │       [build-dependencies]
    │       └── autocfg v1.1.0
    ├── generic-array v0.14.6 (*)
    ├── hkdf v0.12.3
    │   └── hmac v0.12.1
    │       └── digest v0.10.6
    │           ├── block-buffer v0.10.3
    │           │   └── generic-array v0.14.6 (*)
    │           ├── crypto-common v0.1.6
    │           │   ├── generic-array v0.14.6 (*)
    │           │   └── typenum v1.15.0
    │           └── subtle v2.4.1
    ├── num v0.4.0
    │   ├── num-bigint v0.4.3
    │   │   ├── num-integer v0.1.45
    │   │   │   └── num-traits v0.2.15
    │   │   │       [build-dependencies]
    │   │   │       └── autocfg v1.1.0
    │   │   │   [build-dependencies]
    │   │   │   └── autocfg v1.1.0
    │   │   └── num-traits v0.2.15 (*)
    │   │   [build-dependencies]
    │   │   └── autocfg v1.1.0
    │   ├── num-complex v0.4.2
    │   │   └── num-traits v0.2.15 (*)
    │   ├── num-integer v0.1.45 (*)
    │   ├── num-iter v0.1.43
    │   │   ├── num-integer v0.1.45 (*)
    │   │   └── num-traits v0.2.15 (*)
    │   │   [build-dependencies]
    │   │   └── autocfg v1.1.0
    │   ├── num-rational v0.4.1
    │   │   ├── num-bigint v0.4.3 (*)
    │   │   ├── num-integer v0.1.45 (*)
    │   │   └── num-traits v0.2.15 (*)
    │   │   [build-dependencies]
    │   │   └── autocfg v1.1.0
    │   └── num-traits v0.2.15 (*)
    ├── once_cell v1.16.0
    ├── rand v0.8.5
    │   ├── libc v0.2.137
    │   ├── rand_chacha v0.3.1
    │   │   ├── ppv-lite86 v0.2.16
    │   │   └── rand_core v0.6.4
    │   │       └── getrandom v0.2.8
    │   │           ├── cfg-if v1.0.0
    │   │           └── libc v0.2.137
    │   └── rand_core v0.6.4 (*)
    ├── serde v1.0.147
    │   └── serde_derive v1.0.147 (proc-macro)
    │       ├── proc-macro2 v1.0.47 (*)
    │       ├── quote v1.0.21 (*)
    │       └── syn v1.0.103 (*)
    ├── sha2 v0.10.6
    │   ├── cfg-if v1.0.0
    │   ├── cpufeatures v0.2.5
    │   └── digest v0.10.6 (*)
    └── zbus v3.10.0
        ├── async-broadcast v0.5.0
        │   ├── event-listener v2.5.3
        │   ├── futures-core v0.3.25
        │   └── parking_lot v0.12.1
        │       ├── lock_api v0.4.9
        │       │   └── scopeguard v1.1.0
        │       │   [build-dependencies]
        │       │   └── autocfg v1.1.0
        │       └── parking_lot_core v0.9.7
        │           ├── cfg-if v1.0.0
        │           ├── libc v0.2.137
        │           └── smallvec v1.10.0
        ├── async-executor v1.5.0
        │   ├── async-lock v2.6.0
        │   │   ├── event-listener v2.5.3
        │   │   └── futures-lite v1.12.0
        │   │       ├── fastrand v1.8.0
        │   │       ├── futures-core v0.3.25
        │   │       ├── futures-io v0.3.25
        │   │       ├── memchr v2.5.0
        │   │       ├── parking v2.0.0
        │   │       ├── pin-project-lite v0.2.9
        │   │       └── waker-fn v1.1.0
        │   ├── async-task v4.3.0
        │   ├── concurrent-queue v2.1.0
        │   │   └── crossbeam-utils v0.8.14
        │   │       └── cfg-if v1.0.0
        │   ├── fastrand v1.8.0
        │   ├── futures-lite v1.12.0 (*)
        │   └── slab v0.4.7 (*)
        ├── async-io v1.12.0
        │   ├── async-lock v2.6.0 (*)
        │   ├── concurrent-queue v2.1.0 (*)
        │   ├── futures-lite v1.12.0 (*)
        │   ├── libc v0.2.137
        │   ├── log v0.4.17
        │   │   └── cfg-if v1.0.0
        │   ├── parking v2.0.0
        │   ├── polling v2.4.0
        │   │   ├── cfg-if v1.0.0
        │   │   ├── libc v0.2.137
        │   │   └── log v0.4.17 (*)
        │   │   [build-dependencies]
        │   │   └── autocfg v1.1.0
        │   ├── slab v0.4.7 (*)
        │   ├── socket2 v0.4.7
        │   │   └── libc v0.2.137
        │   └── waker-fn v1.1.0
        │   [build-dependencies]
        │   └── autocfg v1.1.0
        ├── async-lock v2.6.0 (*)
        ├── async-recursion v1.0.2 (proc-macro)
        │   ├── proc-macro2 v1.0.47 (*)
        │   ├── quote v1.0.21 (*)
        │   └── syn v1.0.103 (*)
        ├── async-task v4.3.0
        ├── async-trait v0.1.64 (proc-macro)
        │   ├── proc-macro2 v1.0.47 (*)
        │   ├── quote v1.0.21 (*)
        │   └── syn v1.0.103 (*)
        ├── byteorder v1.4.3
        ├── derivative v2.2.0 (proc-macro)
        │   ├── proc-macro2 v1.0.47 (*)
        │   ├── quote v1.0.21 (*)
        │   └── syn v1.0.103 (*)
        ├── dirs v4.0.0
        │   └── dirs-sys v0.3.7
        │       └── libc v0.2.137
        ├── enumflags2 v0.7.5
        │   ├── enumflags2_derive v0.7.4 (proc-macro)
        │   │   ├── proc-macro2 v1.0.47 (*)
        │   │   ├── quote v1.0.21 (*)
        │   │   └── syn v1.0.103 (*)
        │   └── serde v1.0.147 (*)
        ├── event-listener v2.5.3
        ├── futures-core v0.3.25
        ├── futures-sink v0.3.25
        ├── futures-util v0.3.25 (*)
        ├── hex v0.4.3
        ├── nix v0.25.1
        │   ├── bitflags v1.3.2
        │   ├── cfg-if v1.0.0
        │   ├── libc v0.2.137
        │   ├── memoffset v0.6.5
        │   │   [build-dependencies]
        │   │   └── autocfg v1.1.0
        │   └── pin-utils v0.1.0
        │   [build-dependencies]
        │   └── autocfg v1.1.0
        ├── once_cell v1.16.0
        ├── ordered-stream v0.2.0
        │   ├── futures-core v0.3.25
        │   └── pin-project-lite v0.2.9
        ├── rand v0.8.5 (*)
        ├── serde v1.0.147 (*)
        ├── serde_repr v0.1.9 (proc-macro)
        │   ├── proc-macro2 v1.0.47 (*)
        │   ├── quote v1.0.21 (*)
        │   └── syn v1.0.103 (*)
        ├── sha1 v0.10.5
        │   ├── cfg-if v1.0.0
        │   ├── cpufeatures v0.2.5
        │   └── digest v0.10.6 (*)
        ├── static_assertions v1.1.0
        ├── tracing v0.1.37
        │   ├── cfg-if v1.0.0
        │   ├── pin-project-lite v0.2.9
        │   ├── tracing-attributes v0.1.23 (proc-macro)
        │   │   ├── proc-macro2 v1.0.47 (*)
        │   │   ├── quote v1.0.21 (*)
        │   │   └── syn v1.0.103 (*)
        │   └── tracing-core v0.1.30
        │       └── once_cell v1.16.0
        ├── zbus_macros v3.10.0 (proc-macro)
        │   ├── proc-macro-crate v1.2.1
        │   │   ├── once_cell v1.16.0
        │   │   ├── thiserror v1.0.37
        │   │   │   └── thiserror-impl v1.0.37 (proc-macro)
        │   │   │       ├── proc-macro2 v1.0.47 (*)
        │   │   │       ├── quote v1.0.21 (*)
        │   │   │       └── syn v1.0.103 (*)
        │   │   └── toml v0.5.9
        │   │       └── serde v1.0.147
        │   ├── proc-macro2 v1.0.47 (*)
        │   ├── quote v1.0.21 (*)
        │   ├── regex v1.7.1
        │   │   ├── aho-corasick v0.7.20
        │   │   │   └── memchr v2.5.0
        │   │   ├── memchr v2.5.0
        │   │   └── regex-syntax v0.6.28
        │   └── syn v1.0.103 (*)
        ├── zbus_names v2.5.0
        │   ├── serde v1.0.147 (*)
        │   ├── static_assertions v1.1.0
        │   └── zvariant v3.11.0
        │       ├── byteorder v1.4.3
        │       ├── enumflags2 v0.7.5 (*)
        │       ├── libc v0.2.137
        │       ├── serde v1.0.147 (*)
        │       ├── static_assertions v1.1.0
        │       └── zvariant_derive v3.11.0 (proc-macro)
        │           ├── proc-macro-crate v1.2.1 (*)
        │           ├── proc-macro2 v1.0.47 (*)
        │           ├── quote v1.0.21 (*)
        │           └── syn v1.0.103 (*)
        └── zvariant v3.11.0 (*)
```

messense commented:
yeah, I plan to try `linux-keyutils` in maturin as long as it can be used in static musl libc builds.

landhb commented:
Sounds good! Yeah it statically links with musl. I just added the following targets to CI/CD:

```
mips64-unknown-linux-muslabi64
mips-unknown-linux-musl
aarch64-unknown-linux-musl
arm-unknown-linux-musleabihf
i686-unknown-linux-musl
x86_64-unknown-linux-musl
```

brotskydotcom reviewed (changes requested):
Thanks much for the contribution.  But without an issue, I can't really tell what problem you are trying to address with this change:

* Avoiding the build problem with linux-keyutils (already fixed by linux-keyutils).
* Only building one default store on each platform (not accomplished by this PR)
* Making the keyring build grab as few dependencies as possible on any platform so clients can bring just use the superstructure to bring their own store (not accomplished by this PR).

This PR is also not fully backwards compatible with 2.* in terms of API, because the default features no longer build keyring.

Can you please file an issue clarifying your goals and modifying your PR to match those goals?  I'd appreciate it.

messense commented:
Sorry I've moved to use only the `linux-keyutils` feature and dropped `secret-service` from dependency tree, so I don't have any motivation to work on this now.

brotskydotcom commented:
OK, thanks for clarifying.  This has got me thinking about allowing builds with no stores at all, for people who want to bring their own stores (e.g., a remote service).  So maybe I'll file that issue :).

FWIW, wearing my well-worn Python hat, [PyO3](https://github.com/PyO3) looks like a very cool project.
